// LembosMapReduceEnvironmentTest-testCompleteLembosMapReduceEnvironment/index.js
'use strict';

module.exports = {
  config: {},

  jobCleanup: function (job, cb) {
    cb();
  },

  jobSetup: function (job, cb) {
    cb();
  },

  mapCleanup: function (context, cb) {
    cb();
  },

  map: function (key, value, context, cb) {
    cb();
  },

  mapSetup: function (context, cb) {
    cb();
  },

  reduceCleanup: function (context, cb) {
    cb();
  },

  reduce: function (key, values, context, cb) {
    cb();
  },

  reduceSetup: function (context, cb) {
    cb();
  },

  combineCleanup: function (context, cb) {
    cb();
  },

  combine: function (key, values, context, cb) {
    cb();
  },

  combineSetup: function (context, cb) {
    cb();
  },

  partitionCleanup: function (conf, cb) {
    cb();
  },

  partition: function (key, value, numPartitions, cb) {
    cb();
  },

  partitionSetup: function (conf, cb) {
    cb();
  },

  groupCleanup: function (conf, cb) {
    cb();
  },

  group: function (key1, key2, cb) {
    cb();
  },

  groupSetup: function (conf, cb) {
    cb();
  },

  sortCleanup: function (conf, cb) {
    cb();
  },

  sort: function (key1, key2, cb) {
    cb();
  },

  sortSetup: function (conf, cb) {
    cb();
  }
};
